### ISSUE TITLE
getOwnMethod invokes getter function

### ISSUE THREAD

**Issue by capaj** [typescript-rtti/typescript-rtti#52]
> Note: Edited example for `reflect(Foo)` instead of `reflect(Foo.constructor)` --@rezonant

```ts

 class Foo {
       get bar(): number {
        console.log('spy')
      }
    }

const reflected = reflect(Foo)
 const methodRtti = reflected.getOwnMethod('bar') //spy is logged


```

when I use reflect, I would expect that getter is not invoked on the Foo class.
It returns undefined, which is also not good-it should return that `Number`.
It must avoid invoking the getter.
I have not tested this, but I suspect other get methods will have the same problem.

~~This essentially makes it impossible to reflect on getters~~.

EDIT: as pointed out we can reflect getters.

**Comment by chibisuke** [comment]
you're inspecting a property, not a method, so you should use getOwnProperty() instead of getOwnMethod().

**Comment by rezonant** [comment]
This is true, but we should try to avoid inadvertently invoking `get` when possible.

**Comment by rezonant** [comment]
Some interesting findings here:
- This occurs when reflecting on classes without type annotations (ie which have not been compiled using the transformer) by design, though the implication of getter execution hadn't been considered
- It also occurs on classes compiled using the transformer _which have no methods of their own_. This is a bug. The transformer should always emit `rt:m` (the method name list annotation)
- It will invoke _all_ property getters once the first time you try to access any method metadata, not just the getter corresponding to the "method" you are looking for, so there's that

When there is no `rt:m` annotation, the library assumes the class was not compiled with the transformer, so it tries to determine the methods by inspecting the properties of the prototype, hence fetching the value of those properties, and triggering a call for each defined getter. It could be argued whether it _should_ do this- its conceivable that a getter could be destructive, but writing such code would be so... inadvisable... that I'm not sure we should hinder the capabilities of `reflect()` to avoid it. I'm curious if anyone disagrees.

However, the transformer should definitely be emitting `rt:m` even if there are no methods to capture that fact, so that the reflection library doesn't try to infer them.

**Comment by rezonant** [comment]
@capaj could you share more about the use case where you encountered this? Sounds like you might have a use case that might make us want to disable this functionality for unannotated classes, even if we fix the bug so that it doesn't occur for annotated classes.

**Comment by chibisuke** [comment]
> Some interesting findings here:
> 
> * This occurs when reflecting on classes without type annotations (ie which have not been compiled using the transformer) by design, though the implication of getter execution hadn't been considered
> * It also occurs on classes compiled using the transformer _which have no methods of their own_. This is a bug. The transformer should always emit `rt:m` (the method name list annotation)
> * It will invoke _all_ property getters once the first time you try to access any method metadata, not just the getter corresponding to the "method" you are looking for, so there's that
> 
> When there is no `rt:m` annotation, the library assumes the class was not compiled with the transformer, so it tries to determine the methods by inspecting the properties of the prototype, hence fetching the value of those properties, and triggering a call for each defined getter. It could be argued whether it _should_ do this- its conceivable that a getter could be destructive, but writing such code would be so... inadvisable... that I'm not sure we should hinder the capabilities of `reflect()` to avoid it. I'm curious if anyone disagrees.
> 
> However, the transformer should definitely be emitting `rt:m` even if there are no methods to capture that fact, so that the reflection library doesn't try to infer them.

I recently notices another related problem, but didn't have the time to try and make a reproduction case yet. 

I got a parrent class that is not transformed during compile (seperate module), which inclused private fields (starting with a # sign). reflect seems to be either accessing these from the outside, or calling the getter that references them without a propper this argument. I worked around it by exclusing the class by name for now (I don't really need it reflected) but I'll try to reproduce this in a propper testcase when I have some time.

**Comment by rezonant** [comment]
Sounds like you have something like this:

```typescript
class A {
    #foo = 123;
    get foo() { return this.#foo; }
}
```

If so, then yeah that would be this issue. If you `reflect(A)`, it will try to examine the properties on the prototype, which would include `A.prototype.foo`, and it wouldnt have a `this` value at that moment

**Comment by rezonant** [comment]
I suppose the potential dependency on `this` is enough of a reason to change this behavior. I think we should simply detect that the property has defined a getter and skip it in that case. It does mean that some properties that a user considers "methods" won't show up on classes without type annotations, but it's better than crashing, as it would if properties of `this` are referenced

**Comment by rezonant** [comment]
This improvement is now shipped in 0.6.0
